### PR TITLE
Hkatz/update summary

### PR DESCRIPTION
**Pull request by hjkatz**
# Why?

The Summary package wasn't as flexible as needed for the next PR (there's a theme here, lol) for the changes that I'm planning to add for the Web UI and Dashboard packages. (Essentially splitting up the templates and providing different views (summaries) per Namespace, VPAs, etc...).

# What?

3 things are going on in this PR:

1. Reverted #90 (and ensured that the reusability goal behind this PR was still met)
1. Updated `pkg/summary` with new features, functionality, and tests
1. Standardized on the usage of `VPA` over `Vpa` as per golang best practices (this happened in this PR because in reality I had like 30 commits in one PR and I wanted to split them up and this one commit just happened to fit nicely here)

### CODE REVIEW

File: pkg/summary/options.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package summary
+
+import (
+	"github.com/fairwindsops/goldilocks/pkg/kube"
+	"github.com/fairwindsops/goldilocks/pkg/utils"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Functional options
+type Option func(*Options)
```
Comment (rbren) [inline pkg/summary/options.go:10]: Never seen this pattern before. Very cool

Comment (hjkatz) [reply pkg/summary/options.go:10]: I found it a while back and have fallen in love with it. It's just so useful for all sorts of packages and `NewFoo()` functions 😄 Here's a blog post about it if you're interested: https://www.sohamkamani.com/golang/options-pattern/

File: pkg/summary/summary.go [hunk 2]
```diff
@@ -49,119 +58,144 @@ type Summary struct {
...
+		if err != nil {
+			return summary, err
+		}
 	}
 
-	containerExclusions := strings.Split(excludeContainers, ",")
-
-	for _, vpa := range vpas.Items {
-		klog.V(8).Infof("Analyzing vpa: %v", vpa.ObjectMeta.Name)
+	if len(s.vpas) <= 0 {
```
Comment (lucasreed) [inline pkg/summary/summary.go:101]: Does this need to be `<=`? It feels like `== 0` would do the job.

File: pkg/summary/summary.go [hunk 3]
```diff
@@ -49,119 +58,144 @@ type Summary struct {
...
 
-	summary.Namespaces = utils.UniqueString(summary.Namespaces)
+	// get the unique list of namespaces we've seen for this summary
+	summary.Namespaces = summaryNamespaces.List()
+
 	return summary, nil
 }
+
+// Update the list of VPAs that the summarizer uses
+func (s *Summarizer) UpdateVPAs() error {
```
Comment (lucasreed) [inline pkg/summary/summary.go:169]: Does this need to be a public method? It looks like the only caller is in this package.

File: pkg/summary/summary.go [hunk 4]
```diff
@@ -49,119 +58,144 @@ type Summary struct {
...
+	return summarizer
 }
 
-func (client *Client) constructSummary(vpas *v1beta2.VerticalPodAutoscalerList, excludeContainers string) (Summary, error) {
+func (s Summarizer) GetSummary() (Summary, error) {
 	var summary Summary
-	if len(vpas.Items) <= 0 {
-		return summary, nil
+	// cached vpas
+	if s.vpas == nil {
```
Comment (lucasreed) [inline pkg/summary/summary.go:94]: It looks like the dashboard may be the only thing that makes use of the cache, but what happens if new VPAs get created after it creates the initial cache? Will the new ones get picked up? Should there be a TTL of the cache?

Comment (lucasreed) [reply pkg/summary/summary.go:94]: Actually looking closer at the dashboard code I don't think anything will ever cache because a new Summarizer object gets created on every call. See: https://github.com/FairwindsOps/goldilocks/pull/154/files#diff-dc1496f734707825dbb6070e68866784R167

File: pkg/dashboard/dashboard.go [hunk 5]
```diff
@@ -162,7 +164,12 @@ func GetRouter(port int, basePath string, vpaLabels map[string]string, excludeCo
 			return
 		}
 
-		data, err := summary.GetInstance().Run(vpaLabels, excludeContainers)
+		summarizer := summary.NewSummarizer(
```
Comment (lucasreed) [inline pkg/dashboard/dashboard.go:167]: If we want to make use of the caching functionality in the `Summarizer` type, the `Summarizer` object should get passed into `GetRouter` from `cmd/dashboard.go`

File: pkg/summary/options.go (hunk 1 above)
Comment (lucasreed) [reply pkg/summary/options.go:10]: Agreed, I like this a lot!

File: pkg/summary/summary.go [hunk 6]
```diff
@@ -20,13 +20,22 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
```
Comment (lucasreed) [inline pkg/summary/summary.go:23]: Cool, I haven't used this package before. I dig it.

File: pkg/summary/summary.go [hunk 7]
```diff
@@ -49,119 +58,144 @@ type Summary struct {
...
+	// set the cached vpas list directly
+	summarizer.vpas = vpas
 
-	summary, _ := GetInstance().constructSummary(vpas, excludeContainers)
-	return summary, nil
+	return summarizer
 }
 
-func (client *Client) constructSummary(vpas *v1beta2.VerticalPodAutoscalerList, excludeContainers string) (Summary, error) {
+func (s Summarizer) GetSummary() (Summary, error) {
```
Comment (lucasreed) [inline pkg/summary/summary.go:91]: I think this should be a pointer
```suggestion
func (s Summarizer) GetSummary() (Summary, error) {
func (s *Summarizer) GetSummary() (Summary, error) {
```

File: pkg/summary/summary.go [hunk 8]
```diff
@@ -49,119 +58,144 @@ type Summary struct {
...
+		return err
+	}
+	klog.V(10).Infof("Found vpas: %v", vpas)
+
+	s.vpas = vpas
+	return nil
+}
+
+// Run creates a summary of the vpa info for all namespaces.
+func (s Summarizer) listVPAs() ([]v1beta2.VerticalPodAutoscaler, error) {
```
Comment (lucasreed) [inline pkg/summary/summary.go:187]:
```suggestion
func (s *Summarizer) listVPAs() ([]v1beta2.VerticalPodAutoscaler, error) {
```

File: pkg/summary/summary.go (hunk 2 above)
Comment (lucasreed) [reply pkg/summary/summary.go:101]: Or `< 1`

File: pkg/utils/utils.go [hunk 9]
```diff
@@ -15,13 +15,13 @@
 package utils
 
 import (
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (hjkatz) [reply pkg/utils/utils.go:18]: Huh, looks like my auto-import thing in Vim added this for some reason. I'll remove it.

File: cmd/summary.go [hunk 10]
```diff
@@ -36,12 +37,30 @@ func init() {
...
+	Use:   "summary [namespace]",
+	Short: "Generate a summary of the vpa recommendations in a namespace (default: all namespaces).",
+	Long: `Gather all the vpa data in a namespace and generate a summary of the recommendations.
+By default the summary will be about all VPAs in all namespaces.`,
+	Args: cobra.ArbitraryArgs,
 	Run: func(cmd *cobra.Command, args []string) {
+		opts := []summary.Option{}
+
+		// limit to a single namespace
+		if len(args) == 1 {
```
[the remark this answers is not in the document]
Comment (hjkatz) [reply cmd/summary.go:49]: Sure! Added this in.

File: pkg/summary/summary.go [hunk 11]
```diff
@@ -49,119 +58,144 @@ type Summary struct {
...
 
-// Run creates a summary of the vpa info for all namespaces.
-func (client *Client) Run(vpaLabels map[string]string, excludeContainers string) (Summary, error) {
-	klog.V(3).Infof("Looking for VPAs with labels: %v", vpaLabels)
-
-	vpaListOptions := metav1.ListOptions{
-		LabelSelector: labels.Set(vpaLabels).String(),
-	}
+// Returns a Summarizer for a known list of VPAs
+func NewSummarizerForVPAs(vpas []v1beta2.VerticalPodAutoscaler, setters ...Option) *Summarizer {
```
[the remark this answers is not in the document]
Comment (hjkatz) [reply pkg/summary/summary.go:82]: No I don't think so. This is a convenience function for dealing with in memory VPAs and getting information about them without going through the normal builder. I think I wrote it while intending to make use of it and I think the situation just never came up.

File: pkg/summary/summary.go (hunk 2 above)
Comment (hjkatz) [reply pkg/summary/summary.go:101]: haha this is a habit of mine that I use to make sure that my if statements always cover _all_ cases and not just the cases I think could possibly happen. I can change this if y'all prefer a different style.

File: pkg/summary/summary.go (hunk 3 above)
Comment (hjkatz) [reply pkg/summary/summary.go:169]: No, I catch this in an upcoming commit and change this to be private as well as revisit the caching behaviour of the Summarizer object to make a little more sense from a public API perspective.

File: pkg/summary/summary.go (hunk 4 above)
Comment (hjkatz) [reply pkg/summary/summary.go:94]: This is what I think prompted me to change the caching and update behaviour of the Summarizer in the next set of commits where I go through and actually use the API in this package in the web ui and dashboard updates I make.

File: pkg/dashboard/dashboard.go (hunk 5 above)
Comment (hjkatz) [reply pkg/dashboard/dashboard.go:167]: Absolutely! This is a shim for now so that the commits touched a more isolated package at one time. Next up is the dashboard updates (which also includes a revisit to the Summarizer package to assist the dashboard get the information it wants).

File: pkg/summary/summary.go (hunk 7 above)
Comment (hjkatz) [reply pkg/summary/summary.go:91]: Why should this be a pointer receiver if the Summarizer struct shouldn't be modified during the function?

**Comment by hjkatz** [comment]
@sudermanjr Looks like the e2e tests have a period difference in the kubectl responses coming back:

```
Expected: 'No resources found.'
Actual:   'No resources found'
```

Perhaps there's a way to use a regex for this test?

File: pkg/summary/summary.go (hunk 7 above)
Comment (lucasreed) [reply pkg/summary/summary.go:91]: Fair point, I usually like to keep my API of a struct consistent across methods, but you are correct that it doesn't necessarily have to. It may just be an idiosyncrasy of mine that isn't necessary.

The only downside I could think of is if anything in the future was added in this function that would modify the struct, it may not be obvious when things don't work as planned.

File: pkg/summary/summary.go (hunk 8 above)
Comment (lucasreed) [reply pkg/summary/summary.go:187]: I'll leave this and the other discussion above up to you, not a blocker.

File: pkg/summary/summary.go (hunk 2 above)
Comment (lucasreed) [reply pkg/summary/summary.go:101]: Nah it's fine as is. It doesn't seem like it could ever be a negative number, but best not to make assumptions 😄

File: cmd/summary.go [hunk 12]
```diff
@@ -36,12 +37,33 @@ func init() {
...
 
 var summaryCmd = &cobra.Command{
-	Use:   "summary",
-	Short: "Genarate a summary of the vpa recommendations in a namespace.",
-	Long:  `Gather all the vpa data in a namespace and generaate a summary of the recommendations.`,
+	Use:   "summary [namespace]",
+	Short: "Generate a summary of the vpa recommendations in a namespace (default: all namespaces).",
+	Long: `Gather all the vpa data in a namespace and generate a summary of the recommendations.
+By default the summary will be about all VPAs in all namespaces.`,
+	Args: cobra.ArbitraryArgs,
```
Comment (sudermanjr) [inline cmd/summary.go:44]: Curious why you chose to use args instead of a namespace flag here, which would be consistent with other Kube-adjacent CLI tools.

Comment (sudermanjr) [inline cmd/summary.go:44]: Curious why you chose to use args instead of a namespace flag here, which would be consistent with other Kube-adjacent CLI tools.

File: cmd/summary.go [hunk 13]
```diff
@@ -36,12 +37,33 @@ func init() {
...
 	Run: func(cmd *cobra.Command, args []string) {
+		var opts []summary.Option
+
+		// limit to a single namespace
+		if len(args) == 1 {
+			opts = append(opts, summary.ForNamespace(args[0]))
+		} else if len(args) > 1 {
+			// undefined argument behaviour
```
Comment (sudermanjr) [inline cmd/summary.go:52]: I don't think this comment is necessary. The klog.Fatal is information enough here.

File: cmd/summary.go [hunk 14]
```diff
@@ -36,12 +37,33 @@ func init() {
...
+			klog.Fatal("Too many namespace arguments provided!")
+		}
+
+		// exclude containers from the summary
```
Comment (sudermanjr) [inline cmd/summary.go:56]: Same as above, comment is unnecessary IMO

File: pkg/summary/options.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
+package summary
+
+import (
+	"github.com/fairwindsops/goldilocks/pkg/kube"
+	"github.com/fairwindsops/goldilocks/pkg/utils"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Functional options
```
Comment (sudermanjr) [inline pkg/summary/options.go:9]: linter: 

```
pkg/summary/options.go:9:1: comment on exported type Option should be of the form "Option ..." (with optional leading article)
```

There's a few of these. I suggest running `golint ./...` on this PR.

File: pkg/summary/options.go [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+type Option func(*Options)
+
+// internal options for getting and caching the Summarizer's VPAs
+type Options struct {
```
Comment (sudermanjr) [inline pkg/summary/options.go:13]: If these are internal options only, then should the type be un-exported?

File: pkg/summary/summary_test.go [hunk 17]
```diff
@@ -74,8 +57,8 @@ func TestRunSummary(t *testing.T) {
 			Namespace: "testing",
 		},
 		Spec: v1beta2.VerticalPodAutoscalerSpec{
-			TargetRef: &autoscaling.CrossVersionObjectReference{
-				APIVersion: "apps/v1",
+			TargetRef: &autoscalingv1.CrossVersionObjectReference{
+				APIVersion: "extensions/v1beta1",
```
Comment (sudermanjr) [inline pkg/summary/summary_test.go:61]: Why did you change the apiVersion of the deployment in this test? This apiVersion will be completely removed in kube 1.16

**Comment by sudermanjr** [comment]
See #159  for the updates to the e2e

File: pkg/summary/summary.go (hunk 8 above)
Comment (lucasreed) [reply pkg/summary/summary.go:187]: I am actually leaning towards the approach you went with instead. Resolving this. Thanks!

File: pkg/summary/summary.go (hunk 7 above)
Comment (lucasreed) [reply pkg/summary/summary.go:91]: I am actually leaning towards the approach you went with instead. Resolving this. Thanks!

File: cmd/summary.go (hunk 12 above)
Comment (hjkatz) [reply cmd/summary.go:44]: TBH I didn't even think of using a flag here. I was thinking more in terms of how I might want to use the summary command and thought `goldilocks summary <namespace>` made sense to me, but I now think that `goldilocks summary --namespace <namespace>` also makes just as much sense.

Which would you prefer?

File: pkg/summary/options.go (hunk 16 above)
Comment (hjkatz) [reply pkg/summary/options.go:13]: I've changed this to not be exported.

File: pkg/summary/options.go (hunk 15 above)
Comment (hjkatz) [reply pkg/summary/options.go:9]: I've fixed up the golint suggestions.

File: pkg/summary/summary_test.go (hunk 17 above)
Comment (hjkatz) [reply pkg/summary/summary_test.go]: I'm not exactly sure. I've changed this to "apps/v1".

File: cmd/summary.go (hunk 12 above)
Comment (sudermanjr) [reply cmd/summary.go:44]: I generally lean towards a flag myself.

File: pkg/summary/options.go [hunk 18]
```diff
@@ -6,21 +6,20 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 )
 
-// Functional options
-type Option func(*Options)
+type option func(*options)
```
Comment (lucasreed) [inline pkg/summary/options.go:9]: This should not have been unexported. It's used in the `cmd` package. Once it's exported again you'll need to update the references in the the `summary` package as well.

Comment (hjkatz) [reply pkg/summary/options.go]: Whoops! Nice catch, sorry for that!

File: cmd/summary.go (hunk 12 above)
Comment (hjkatz) [reply cmd/summary.go:46]: 👍

**Comment by hjkatz** [comment]
So I think github is stuck or something. It looks like the PR thing says there's 1 change requested by @lucasreed, but I can't find the request or the review. Do y'all have any ideas on what changes are still wanted by Github?

> Merging can be performed automatically once the requested changes are addressed.

**Comment by lucasreed** [comment]
> So I think github is stuck or something. It looks like the PR thing says there's 1 change requested by @lucasreed, but I can't find the request or the review. Do y'all have any ideas on what changes are still wanted by Github?
> 
> > Merging can be performed automatically once the requested changes are addressed.

I clicked requested changes when re-reviewing for the `Options` type being unexported. I just needed to review again.
